packages/access-api/src/service/upload-api-proxy.js

```diff
@@ -0,0 +1,188 @@
+  /**
+   * @type {UploadService}
+   */
+  const store = {
+    add: handleInvocation,
+    list: handleInvocation,
+    remove: handleInvocation,
+  }
+  return store
+}
+
+/**
+ * @typedef UcantoHttpConnectionOptions
+ * @property {dagUcan.DID} audience
+ * @property {typeof globalThis.fetch} options.fetch
+ * @property {URL} options.url
+ */
+
+/**
+ * @param {UcantoHttpConnectionOptions} options
+ * @returns {Ucanto.ConnectionView<any>}
+ */
+function createUcantoHttpConnection(options) {
+  return Client.connect({
+    id: DID.parse(options.audience),
+    encoder: CAR,
+    decoder: CBOR,
+    channel: HTTP.open({
+      fetch: options.fetch,
+      url: options.url,
+    }),
+  })
+}
+
+/**
+ * @type {Record<string, {
+ * audience: dagUcan.DID,
+ * url: URL,
+ * }>}
+ */
+const uploadApiEnvironments = {
+  production: {
+    audience: 'did:web:web3.storage',
+    url: new URL('https://up.web3.storage'),
+  },
+  staging: {
+    audience: 'did:web:staging.web3.storage',
+    url: new URL('https://staging.up.web3.storage'),
+  },
+}
+
+/**
+ * @interface {Pick<Map<dagUcan.DID, Ucanto.Connection<any>>, 'get'>}
+ */
+const audienceConnections = {
+  audienceToUrl: (() => {
+    /** @type {{ [k: keyof typeof uploadApiEnvironments]: URL }} */
+    const object = {}
+    for (const [, { audience, url }] of Object.entries(uploadApiEnvironments)) {
+      object[audience] = url
+    }
+    return object
+  })(),
+  fetch: globalThis.fetch,
+  /** @type {undefined|Ucanto.DID} */
+  defaultAudience: uploadApiEnvironments.production.audience,
+  /**
+   * Return a ucanto connection to use for the provided invocation audience.
+   * If no connection is available for the provided audience, return a connection for the default audience.
+   *
+   * @param {dagUcan.DID} audience
+   */
+  get(audience) {
+    const defaultedAudience =
+      audience in this.audienceToUrl ? audience : this.defaultAudience
+    if (!defaultedAudience) {
+      return
+    }
+    const url = this.audienceToUrl[defaultedAudience]
+    return createUcantoHttpConnection({
+      audience: defaultedAudience,
+      fetch: this.fetch,
+      url,
+    })
+  },
+}
+
+export class UploadApiProxyService {
+  /** @type {StoreService} */
+  store
+  /** @type {UploadService} */
+  upload
+
+  /**
+   * @param {object} options
+   * @param {Ucanto.Signer} [options.signer]
+   * @param {typeof globalThis.fetch} options.fetch
+   */
+  static create(options) {
+    const proxyOptions = {
+      signer: options.signer,
+      connections: {
+        ...audienceConnections,
+        ...(options.fetch && { fetch: options.fetch }),
+      },
+    }
+    return new this(
+      createProxyStoreService(proxyOptions),
+      createProxyUploadService(proxyOptions)
+    )
+  }
+
+  /**
+   * @protected
+   * @param {StoreService} store
+   * @param {UploadService} upload
+   */
+  constructor(store, upload) {
+    this.store = store
+    this.upload = upload
+  }
+}
```

Comment:
having an hard time following this

cant we just to the following ? all the types should match and its easier to understand
```js

/**
 * @param {object} options
 * @param {typeof globalThis.fetch} options.fetch
 */
function createUploadProxy(options) {
  /** @type {Record<Ucanto.DID, Ucanto.ConnectionView<import('@web3-storage/access/types').Service>>} */
  const connections = {
      'did:web:web3.storage': createUcantoHttpConnection({
        audience: 'did:web:web3.storage',
        url: new URL('https://up.web3.storage'),
        fetch: options.fetch
      }),
     'did:web:staging.web3.storage': createUcantoHttpConnection({
        audience: 'did:web:staging.web3.storage',
        url: new URL('https://staging.up.web3.storage'),
        fetch: options.fetch
      })
    }
  

  return {
    store: {
      add: Server.provide(Store.add, async ({capability, invocation}) => {
        const conn = connections[invocation.audience.did()]
        return conn.execute(invocation)
      })
    },
    .... // all the other handlers
  }
}
```

Reply:
I iterated towards this in two ways:
* `connections` is now just an object like you illustrate here no `get` method
* unbundling the class with store + upload into separate functions that build each

I don't believe that the way you're doing `conn.execute(invocation)` will actually work. I wish it did. I tried that first. but had to try some various things to avoid making ucanto changes.

@Gozala and I discussed a few ways that ucanto could make this easier, but also decided to do something that works with ucanto as is, ship that, and decde after that whether to push some of those tools (that are now in `access-api/src/ucanto/proxy.js`) down into ucanto itself.

Reply:
why does it not work ? this was the promise of "lets make a single endpoint because ucanto makes this super simple" this is not super simple